File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] answer = new int[n];
        answer[0] = 1;

        int leftmul = 1;
        for (int i = 0; i < n - 1; i++) {
            leftmul *= nums[i];
            answer[i+1] = leftmul;
        }
        int rightmul = 1;
        for (int i = n - 1; i > 0 ; i--) {
            rightmul *= nums[i];
            answer[i-1] *= rightmul;
        }

        return answer;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        prods = [1]
        for i in range(len(nums) - 1):
            prods.append(nums[i] * prods[-1])
        multiplier = 1
        for i in range(1, len(nums) + 1):
            prods[-i] *= multiplier
            multiplier *= nums[-i]
        return prods

        